### DISCUSSION TITLE
A `RuntimeError: Stream consumed` error is raised when `request.body()` is read in the custom middleware

### DISCUSSION THREAD

**Discussion by theredfoxlee** [encode/starlette#2556] [Potential Issue] [answered]
Hi,

I encountered the following issue while using Starlette today :)

## Issue description

A `RuntimeError: Stream consumed` error is raised when `request.body()` is read in the custom middleware (1) after `call_next`, given that a route uses `request.body()`.

### Code

```python
from starlette.applications import Starlette
from starlette.responses import JSONResponse
from starlette.routing import Route
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.middleware import Middleware

async def testpage(request):
    await request.body()
    return JSONResponse({'hello': 'world'})


class CustomMiddleware(BaseHTTPMiddleware):

    async def dispatch(self, request, call_next):
        response = await call_next(request)
        await request.body()  # THIS WILL RAISE -> RuntimeError: Stream consumed
        return response


app = Starlette(
    debug=True,
    routes=[Route('/', testpage, methods=["PUT"])],
    middleware=[Middleware(CustomMiddleware)]
)
```

### Dependencies

```txt
starlette==0.37.2
uvicorn==0.29.0
```

### Commands

- 1st terminal: `uvicorn example:app`
- 2nd terminal: `curl -XPUT http://127.0.0.1:8000 -d '123'`

## Known workarounds

1. Use `request.body()` at least once before `call_next` to ensure that the `_body` is cached.

> I prefer not to use a workaround, as I only want to read the body in the middleware conditionally, based on the response, rather than always.

---

This is my first time using `starlette`. Is this usage incorrect, or is it a bug?

**Comment by theredfoxlee** [comment]
I believe it's a bug. May I move it to the issues?

**Reply by Kludex** [reply]
No.

**Reply by theredfoxlee** [reply]
Oh... what does it take to state that this is an issue, then? :thinking:

**Reply by Kludex** [reply]
A maintainer to label this as an issue.

This is probably not an issue, you didn't check the discussions, and issues about the topic.

**Reply by lukasab** [reply]
@theredfoxlee checkout https://github.com/getsentry/sentry-python/issues/1675#issuecomment-1675287630

**Reply by theredfoxlee** [reply]
@lukasab, thank you for your reply. However, I believe this situation is different as there is no exception in the simple middleware I presented above, so there's no reason to check whether the request is disconnected or not. **The problem here is that Starlette does not allow reading the request.body in the middleware after the route has read the request.body**. On the other hand, it does allow for reading it before the route reads the request body. Is this expected behavior? I have not seen it documented anywhere. I would expect that the body would be cached once read by the route.

**Reply by theredfoxlee** [reply]
@Kludex, isn't the reproducer posted above enough to state that it's a bug or misuse? The situation occurs in 100% of situations

**Reply by lukasab** [reply]
I don't get it. Isn't your middleware raising the Runtime exception? This means that during the middleware the client has disconnected which internally raises the `RuntimeError` with "Stream consumed" when you try to read the request body. Would this work for you?

```python
class CustomMiddleware(BaseHTTPMiddleware):

    async def dispatch(self, request, call_next):
        response = await call_next(request)
        if await request.is_disconnected():
            return Response(status_code=HTTPStatus.NO_CONTENT)
        await request.body() # should not raise the error anymore
        return response
```

**Reply by theredfoxlee** [reply]
> Isn't your middleware raising the Runtime exception?

It is.

> This means that during the middleware the client has disconnected

Sadly, it doesn't always mean that.

`await request.is_disconnected()` returns `False` in this situation, which means that the client is connected and `request.body()` still raises the exception despite of it (due to its internal implementation issues).

Here's the same exception with your snippet: [log.txt](https://github.com/user-attachments/files/15513256/log.txt).

---

Once again, **I think it's a bug**. Starlette already implemented a workaround for reading the body before `call_next` last year: https://github.com/encode/starlette/pull/1692; what's left is handling the same situation after `call_next`.

**Reply by theredfoxlee** [reply]
The issue here is that we're using `starlette.requests.Request` inside the route, which seems not to propagate the read `.body` to `starlette.middleware.base._CachedRequest` used in the middleware.

**Reply by ex3cv** [reply]
Hi @theredfoxlee

Just move `await request.body()` to be called first in your method. That should do the trick. ;-)
```
async def dispatch(self, request, call_next):
    await request.body()
    response = await call_next(request)
    return response
```

**Reply by theredfoxlee** [reply]
Hi @ex3cv, thanks, but I've already mentioned that in _Known workarounds_. It doesn't solve the bug; it avoids it :/

**Reply by adriangb** [reply, marked as answer]
This isn't a bug, this is just how things work with ASGI / streaming data.

If you don't want to first read (and thus cache) the request body in the middleware I recommend you do something like this:

```python
from random import randint

from starlette.applications import Starlette
from starlette.middleware import Middleware
from starlette.middleware.base import BaseHTTPMiddleware, RequestResponseEndpoint
from starlette.requests import Request
from starlette.responses import Response
from starlette.routing import Route


async def read_body_cached(request: Request) -> bytes:
    request.scope['cached_body'] = body = await request.body()
    return body


async def endpoint(request: Request) -> Response:
    if randint(1, 1) > 0.5:  # some condition you want
        body = await read_body_cached(request)
    else:
        body = await request.body()
    return Response(body)


async def middleware(request: Request, call_next: RequestResponseEndpoint) -> Response:
    response = await call_next(request)
    body: bytes | None = request.scope.get('cached_body')
    if body:
        response.headers['X-Cached-Body'] = body.decode()
    return response


app = Starlette(
    debug=True, routes=[Route('/', endpoint, methods=['PUT'])], middleware=[Middleware(BaseHTTPMiddleware, dispatch=middleware)]
)

if __name__ == '__main__':
    from starlette.testclient import TestClient

    client = TestClient(app)

    response = client.put('/', content=b'Hello, world!')
    assert response.status_code == 200
    assert response.text == 'Hello, world!'
    assert response.headers['X-Cached-Body'] == 'Hello, world!'
```

This assumes that you know that there is no middleware that mutates the body (e.g. decompression), etc.
